### PR TITLE
Fix: Pre save hook modified to run only once during registration

### LINKED ISSUE #41: [Bug]: Better Implementation required for Auth Module
### What happened?
- Password validation fails to match encrypted and plain password entered by user 
- Either it allows anyone to pass the validation and get access or it allows access to none 
- Current Implementation using `bcrypt` is not full proof, since it is not that efficient. Also it can be decrypted. 
> Any Good Implementation which is fool-proof is welcomed. Can be discussed in comments here and finalised

### How can we reproduce this bug?
- A new account was created, and the password was carefully entered during registration.
- Account Verification was done. After Verification, when Password was entered - The user was denied access on `Correct Password`

- In Another Implementation, user was given access on `Wrong Password`

### Desktop Information (Optional)
Operating System: Windows
Browser: Brave
Node Version: 16.15.1

### Urgency (Optional)
High priority

### PR DESCRIPTION
**Pull request by NimishKashyap**
### 🛠️ Fixes Issue
Closes #41

### 🔧 Check List (Check the type of issue it fixes) 

- [x] Bug Fix
- [ ] New Extension
- [ ] Enhancement
- [ ] Documentation 

### ✔️ Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have done `git pull` before adding my new changes
- [x] The title of my pull request is a short description of the requested changes.

### 📄 Note to reviewers
Pre-save hook of mongoose is modified to run the hashing logic of password only once during registration by adding hookEnabled hook into User Model.

Review and testing required.

### CODE REVIEW

File: routes/api.js [hunk 1]
```diff
@@ -42,7 +48,56 @@ const api = (app) => {
...
+                .isEmpty()
+                .withMessage("name property should exist"),
+            body("description")
+                .not()
+                .isEmpty()
+                .withMessage("description should exist"),
+            body("department").custom((value) => {
+                let name = value.toLowerCase();
+                if (
+                    name !== "programming" &&
```
Comment (hariketsheth) [inline routes/api.js:67]: Do we require this in the backend for Departments?
The option given on the frontend anyways is a type of <List> , So values can never be anything else apart from [electrical, programming, electrical, robotics]

**Review by hariketsheth — APPROVED** [review]
Rest, LGTM 👍🏻 
Tested out Authentication, it works well now ! 🎉

**Comment by hariketsheth** [comment]
@NimishKashyap @prateekmohanty63 
Should we still keep #41 open for any other better implementation other than using `bcrypt`. 
Or its fine as of now we will mark it as resolved.

File: routes/api.js (hunk 1 above)
Comment (NimishKashyap) [reply routes/api.js:67]: Actually I am putting this check as confirmation, as our api will be kind of public as of now and has CORS enabled, anyone will be able to send API request to our app using postman, hence I am putting this check as validation.

File: routes/api.js [hunk 2]
```diff
@@ -42,7 +48,56 @@ const api = (app) => {
...
+                if (given <= current) {
+                    return Promise.reject(
+                        "Start time should not be less than current time"
+                    );
+                }
+                return Promise.resolve();
+            }),
+            body("end").custom((value, { req }) => {
+                let given = new Date(value).getTime();
+                if (given <= new Date(req.body.start).getTime()) {
```
[the remark this answers is not in the document]
Comment (NimishKashyap) [reply routes/api.js:90]: Well again, this is backend validation is for added security.

**Comment by NimishKashyap** [comment]
> @NimishKashyap @prateekmohanty63 Should we still keep #41 open for any other better implementation other than using `bcrypt`. Or its fine as of now we will mark it as resolved.

Do we need other implementation than 'bcrypt'? I used 'bcrypt' because of CodeQL error that I was facing with normal crypto nodejs package.

File: routes/api.js (hunk 1 above)
Comment (hariketsheth) [reply routes/api.js:67]: Yep, that makes sense. 
Sure then this looks good 👍🏻

File: routes/api.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hariketsheth) [reply routes/api.js:90]: Yep 👍🏻

**Comment by hariketsheth** [comment]
> > @NimishKashyap @prateekmohanty63 Should we still keep #41 open for any other better implementation other than using `bcrypt`. Or its fine as of now we will mark it as resolved.
> 
> Do we need other implementation than 'bcrypt'? I used 'bcrypt' because of CodeQL error that I was facing with normal crypto nodejs package.

Uhmm, I meant if any other better service is available in comparison to `bcrypt`

**Comment by prateekmohanty63** [comment]
I think we can close the issue as of now
And maybe after others are solved
We can again bring it for s better implementation

Or reducing the priority also works

**Comment by NimishKashyap** [comment]
Yes than we can merge this pull request.

**Pull request merged**